src/voice/gateway/backends/tungstenite.rs

```diff
@@ -0,0 +1,69 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use futures_util::{
+    stream::{SplitSink, SplitStream},
+    StreamExt,
+};
+use tokio::net::TcpStream;
+use tokio_tungstenite::{
+    connect_async_tls_with_config, tungstenite, Connector, MaybeTlsStream, WebSocketStream,
+};
+
+use crate::{errors::VoiceGatewayError, voice::gateway::VoiceGatewayMessage};
+
+#[derive(Debug, Clone)]
+pub struct TungsteniteBackend;
+
+// These could be made into inherent associated types when that's stabilized
+pub type TungsteniteSink =
+    SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, tungstenite::Message>;
+pub type TungsteniteStream = SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>;
+
+impl TungsteniteBackend {
+    pub async fn connect(
+        websocket_url: &str,
+    ) -> Result<(TungsteniteSink, TungsteniteStream), crate::errors::VoiceGatewayError> {
+        let mut roots = rustls::RootCertStore::empty();
+        for cert in rustls_native_certs::load_native_certs().expect("could not load platform certs")
```

Comment:
Can this loading of certs fail in any reasonable (especially outside-influenced) circumstance? If so, `.expect()` should be `?`. If `load_native_certs()` doesn't usually fail, `.expect()` is fine.

Reply:
It apparently can in some platform specific ways, I've added a quick check along with returning `VoiceGatewayError::CannotConnect`